### PR TITLE
fix(highlight): add force in nvim_set_hl

### LINKED ISSUE #20323: `nvim_set_hl` is missing the `!` behavior of `hi! def link`
### Neovim version (nvim -v)

NVIM v0.7.2

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Arch Linux on Windows 10 x86_64

### Terminal name/version

windows terminal

### $TERM environment variable

xterm-256color

### Installation

system package manager : pacman

### 1. 
nvim --clean debug.vim

debug.vim:

```vim
" This is a vim comment.
```
### 2. For `hi! def link`

```vim
:hi vimComment
vimComment     xxx links to Comment

:hi! def link vimComment Boolean

:hi vimComment
vimComment     xxx links to Boolean
```

### 3. For `nvim_set_hl`

```vim
:hi vimComment
vimComment     xxx links to Comment

:call nvim_set_hl(0, 'vimComment', {'link':'Boolean','default':1})

:hi vimComment
vimComment     xxx links to Comment
```

### Expected behavior

Expected an options to make `nvim_set_hl` behave the same as `hi! def link`.

### Actual behavior

If `vimComment` already has highlight, `:hi! def link vimComment Boolean` would override the origin highlight, but `:call nvim_set_hl(0, 'vimComment', {'link':'Boolean','default':1})` can't override it.

### LINKED ISSUE #21119: Function to modify already set highlight groups
### Feature already in Vim?

Not that I know of

### Feature description

In the lua api there is currently the function `nvim_set_hl()` which sets a highlight group, it also overrides the highlight group that you specify if it already exists. The `:highlight` command is able to modify and not override a highlight group. Could there be a lua function that allows you to modify but not override a highlight group?

I have looked in the open and closed Issues and I can't find anything similar to this.

### PR DESCRIPTION
**Pull request by glepnir**
add `force` option and make `nvim_set_hl` works like `highlight` command

Fix #21119

### CODE REVIEW

File: src/nvim/highlight_group.c [hunk 1]
```diff
@@ -799,31 +799,41 @@ int lookup_color(const int idx, const bool foreground, TriState *const boldp)
...
       g->sg_deflink = link_id;
       g->sg_deflink_sctx = current_sctx;
       g->sg_deflink_sctx.sc_lnum += SOURCING_LNUM;
     }
+    // Don't allow a link when there already is some highlighting
+    // for the group, unless force set
+    if (!dict->force && has_set) {
+      if (SOURCING_NAME == NULL && !is_default) {
+        emsg(_(e_group_has_settings_highlight_link_ignored));
+      }
```
Comment (bfredl) [inline src/nvim/highlight_group.c:819]: This part should not be done. It is a _deliberate_ difference of the API that links are not somehow less worth than filling in attributes.

Comment (glepnir) [reply src/nvim/highlight_group.c:819]: okay will remove this part.

Comment (glepnir) [reply src/nvim/highlight_group.c]: how about now :thinking:

File: src/nvim/highlight_group.c [hunk 2]
```diff
@@ -799,31 +799,37 @@ int lookup_color(const int idx, const bool foreground, TriState *const boldp)
 void set_hl_group(int id, HlAttrs attrs, Dict(highlight) *dict, int link_id)
 {
   int idx = id - 1;  // Index is ID minus one.
-
   bool is_default = attrs.rgb_ae_attr & HL_DEFAULT;
-
-  // Return if "default" was used and the group already has settings
-  if (is_default && hl_has_settings(idx, true)) {
```
Comment (bfredl) [inline src/nvim/highlight_group.c:806]: isn't it enough to just check if `force` is set here and then override this check right here? I don't think making the code more convoluted below is needed or desired.

Comment (glepnir) [reply src/nvim/highlight_group.c:806]: before tried but test case failed.

Comment (glepnir) [reply src/nvim/highlight_group.c:806]: these will failed if move before link set the `default link` can't set .  test case from issue 23023 reproduce step
 
```lua
    meths.set_hl(0, 'Bar', {link ='Boolean', default = true })
    eq('Bar            xxx guifg=#123456',
      exec_capture('highlight Bar'))
    command('hi clear')
    eq('Bar            xxx links to Boolean',
      exec_capture('highlight Bar'))
```

File: src/nvim/api/vim.c [hunk 3]
```diff
@@ -151,6 +151,7 @@ Dictionary nvim_get_hl(Integer ns_id, Dict(get_highlight) *opts, Arena *arena, E
 ///                - cterm: cterm attribute map, like |highlight-args|. If not set,
 ///                         cterm attributes will match those from the attribute map
 ///                         documented above.
+///                - force: boolean if true force update the highlight group when it exist.
```
Comment (bfredl) [inline src/nvim/api/vim.c:154]: `boolean` is redundant if we say "if true". also "when it exists" to be inglish grammar.

Comment (glepnir) [reply src/nvim/api/vim.c]: get it.

**Comment by fitrh** [comment]
Maybe I'm misunderstanding, but this still doesn't behave like `:highlight`, for example with the following config

```lua
vim.o.termguicolors = true
vim.cmd.colorscheme("quiet")
```


Using `:higlight` gives the following results

```vim
:hi DiagnosticUnderlineError
DiagnosticUnderlineError xxx cterm=underline gui=underline guisp=Red
:hi DiagnosticUnderlineError gui=undercurl
:hi DiagnosticUnderlineError
DiagnosticUnderlineError xxx cterm=underline gui=undercurl guisp=Red
```

Meanwhile `nvim_set_hl` is as follows

```vim

:hi DiagnosticUnderlineError
DiagnosticUnderlineError xxx cterm=underline gui=underline guisp=Red
:lua vim.api.nvim_set_hl(0, 'DiagnosticUnderlineError', { undercurl = true })
:hi DiagnosticUnderlineError
DiagnosticUnderlineError xxx cterm=undercurl gui=undercurl
```

**Comment by zeertzjq** [comment]
Maybe it's better to name this `keep` rather than `force`?

**Comment by zeertzjq** [comment]
It seems to only work with `default = true`?

**Comment by clason** [comment]
Yes, there was a very specific scenario which didn't work as expected. It is _expected_ (and intended) that `nvim_set_hl` (without options) behaves differently from `:highlight` (without modifiers).

**Comment by glepnir** [comment]
should check `-1` avoid before value overwrite.

**Pull request merged**